V2

SaculRennorb opened the pull request:
Rewritten significant parts of the whole tool, all parsers work properly now.

The focus of this second iteration was primarily to fully extract all information form any mod and improve error reporting as well as the validation itself.

The architecture is also set up in a way where is should be easy to incorporate into the game in the future, and integrate with any error reporting there. The new system tries to parse as much as possible, while reporting errors along the way. After that a validation step can be called, to remove any invalid values and report issues with those.

Issue reporting works via a callback, that gets feed errors as they happen. Errors are represented by a tagged enum that can be processed as desired. That way it is also possible to easily provide localization for errors, as s likely required by the game. Instead of the tagged union we could also use Exception supertyping (not throwing them, just passing them to the callback), I just wanted to try out the current approach.


I've also taken the liberty to convert the project file itself to a more recent version, that allows specifying all the application properties in one place.
This Now also includes a publish profile (just folder publishing) that is correctly set up to produce an assembly for use on the moddb server. 


Oh, also this now has some tests.
They use actual mods, but I've stripped out the assets from the zips as those are not of interest to us and only bloat the repository size.
I haven't bothers setting up GH runners for the tests, as realistically this is not worth the effort. 

It is not required to copy all the dependencies that get emitted from including the cs parser, if found that only

Microsoft.CodeAnalysis.CSharp.dll
Microsoft.CodeAnalysis.dll
System.Collections.Immutable.dll
System.Runtime.CompilerServices.Unsafe.dll
System.Memory.dll

are required.

Once we move to dotnet8 we can also just build this as a single executable, avoiding the clutter.


Example outputs of the "new" tool:

```
$ ModPeek.exe -h 

ModPeek.exe - extract information from VintageStory mod files:

Synopsis:
        ModPeek.exe [-i] [-p] [-f] input

Options:
        -f file, --file file:
                The mod file to operate on.
                If this option is not specified the last standalone argument will be used instead.

        -i, --idandversion
                Only print   modid:version   and exit.

        -p, --always-print
                Print partial info even if errors occurred (if possible).
                If this is not set any errors will suppress normal output.

Operands:
        input
                The mod file to extract information from.
```

```
$ ModPeek.exe StepUp-v1.2.0.cs 
Id: stepup
Name: StepUp
Version: 1.2.0
NetworkVersion:
Dependencies: game
Authors: copygirl
Contributors:
Website: https://www.vintagestory.at/forums/topic/3349-stepup-v120/
Description: Doubles players' step height to allow stepping up full blocks
```


```
$ModPeek.exe "TexturePack and CloneTool 1.0.4.zip"
Unexpected property 'dependency' with value '{"game":"1.20.9"}'.
```


```
$ModPeek.exe warriordrink_v1.0.2.zip                
Unexpected property 'author' with value '"Flint_N_Steel"'.
Unexpected property 'gameversions' with value '["1.19.0"]'.
```

SaculRennorb committed:
Rewritten significant parts of the whole tool, all parsers wrok properly now.

The focus of this second iteration was primarily to fully extract all information form any mod.

The architecure is also set up in a way where is should be easy to incoperate into the game in the future, and integrate with any error reporting there.
The ny system tries to parse as much as possible, while reproting errors along the way.
After that a validation step can be called, to remove any invalid values and report issues with thos.

Issue reporting works via a callback, that gets fead errors as tehy happen.
Errors are represented by a tagged enum that can be processed as desired. That way it is also possible to easily provide localization for errors, as s liekly required by the game.
Instead of the tagged union we could also use Exception supertying (not throwing them, just passing them to the callback), I just wanted to try out the current approach.

Oh, also this now has some tests.
They use actual mods, but ive stripped aut the assets from the zips as those are not of interest to us and only bloat the repository size.

It is not required to copy all the dependencies that get emmited from including the cs parser, if found that only

Microsoft.CodeAnalysis.CSharp.dll
Microsoft.CodeAnalysis.dll
System.Collections.Immutable.dll
System.Runtime.CompilerServices.Unsafe.dll
System.Memory.dll

are required.

Once we move to dotnet8 we can also just build this as a single executable, avoiding the clutter.

SaculRennorb committed:
removed old leftover comemnts

SaculRennorb committed:
ouput lowercase binary like before

SaculRennorb committed:
added test for valid dll

tyronx commented on ModPeek/ParseZIP.cs:
```diff
@@ -0,0 +1,277 @@
+        var error = false;
+        modInfo = new ModInfo();
+
+        foreach(var prop in ((JObject)root).Properties()) {
```
can we avoid code duplication? (nearly the same code in ParseCS.cs)

SaculRennorb replied:
Well, they are similar, but one parses a json and the other parses an assembly attribute ast.
These are different structures, process completely different types and produce different errors, so I don't they ca reasonably be de-duplicated.

SaculRennorb committed:
"Simpler" error structure

The previous iteration of the Error struct was a good experiment, but its overly complicated for what it does and simply fihgts the language on several fronts for no good reason.

SaculRennorb committed:
dont exit after parser error if the always-print option is set

SaculRennorb committed:
Added escaping for reserved delimiter in author and contributor names

SaculRennorb commented:
manually merged
